_posts/2020-09-30-ghelamco-alert.md

```diff
@@ -0,0 +1,824 @@
+The broker is a server that receives all messages from the clients and then routes those messages to relevant destination clients.  
+A client is anything that can interact with the broker to send and receive messages.  
+Our client is the RPI but it could also be an IoT sensor in the field or an application in a data center that processes IoT data.  
+
+In our backend java code we have an MQTTJobService which connects to AWS IoT by using the [AWS SDK](https://aws.amazon.com/sdk-for-java/) and subscribes to the relevant topics to receive and respond to AWS Iot jobs.  
+Every 30 seconds we will read these topics to see if there are any new jobs to be processed.  
+
+### AWS IOT Greengrass
+
+AWS greengrass is a service that extends the AWS cloud onto your edge device.  
+This was fascinating for us since we had some tough problems to solve:  
+* How can we deploy our application on the RPI device?  
+* How can we make sure our application recovers from failures?
+* How can we keep our system itself up to date?
+* How do we find the network address from our device when we are not in the same network?
+
+AWS greengrass offers solutions to all these challenges!  
+Greengrass is an extension of the AWS IoT service and since we had already set up our device in AWS IoT it was easy for us to set up greengrass.  
+To get started with greengrass we had to do 2 additional steps:
+* Define a greengrass group. This group will contain your IoT devices and deployments.  
+* Define a greengrass core. The core is the device that you will use to run the additional AWS capabilities on.  
+
+The greengrass group allows us to push deployments to all the devices in our group.  
+It also allows you to run lambda functions on your core device or install additional connectors with your own runtime of choice.  
+<div style="text-align: center;">
+  <img alt="Greengrass group" src="/img/2020-09-25-ghelamco-alert/gg_group.PNG" width="auto" height="auto" target="_blank" class="image fit">
+</div>
+
+When you register your device as a core device in greengrass you immediately get some nice additional benefits from this.  
+For example, you can immediately see all the network interfaces on your core device and what ip addresses got allocated.  
+This is especially useful if you want to ssh to your device and do not have fixed ip attached to it.  
+<div style="text-align: center;">
+  <img alt="Greengrass core" src="/img/2020-09-25-ghelamco-alert/gg_core.PNG" width="auto" height="auto" target="_blank" class="image fit">
+</div>
+
+Now we that greengrass is installed and our RPI is configured as a core device our greengrass group we can start making full use of the capabilities that greengrass offers.  
+We decided that we wanted our application to run as a docker container, so we installed and configured the **docker connector** for greengrass.   
```

Comment:
Add a link to that docker connector page

like: https://docs.aws.amazon.com/greengrass/latest/developerguide/docker-app-connector.html